astropy/io/fits/tests/test_connect.py

```diff
@@ -304,6 +305,42 @@ def check_equal(filename, expected, start_from=1):
         check_equal(filename, 3, start_from=2)
         assert equal_data(t2, Table.read(filename, hdu=1))
 
+    def test_inexact_format_parse_on_read(self, tmpdir):
+        filename = str(tmpdir.join('test_inexact_format_parse_on_read.fits'))
+        c1 = fits.Column(name='a', array=np.array([1, 2, np.nan]), format='E')
+        table_hdu = fits.TableHDU.from_columns([c1])
+        table_hdu.writeto(filename)
+
+        with fits.open(filename) as hdulist:
+            assert read_table_fits(hdulist)
+
+            data = hdulist[1].data
+
+            for col in data.columns:
+                with pytest.raises(TypeError):
+                    assert np.isnan(col.array)
```

Comment:
I think the test could be simpler, e.g. `Table.read(...)['col0'].mask == [False, False, True]`.
This file is testing the fits/Tabe interface, so generally tests should just use the Table interface.